src/events/attributes.rs

```diff
@@ -4,13 +4,16 @@
 
 use crate::errors::{Error, Result as XmlResult};
 use crate::escape::{do_unescape, escape};
+use crate::escapei::{self, EscapeError};
 use crate::name::QName;
 use crate::reader::{is_whitespace, Reader};
 use crate::utils::{write_byte_string, write_cow_string, Bytes};
 use std::fmt::{self, Debug, Display, Formatter};
 use std::iter::FusedIterator;
 use std::{borrow::Cow, collections::HashMap, io::BufRead, ops::Range};
 
+type CustomEntityMap = HashMap<Vec<u8>, Vec<u8>>;
```

Comment:
I propose to step further and change the API, because I'm unpleasant with the current one:
- keys and values should be UTF-8 strings (as stated in one functions' comment), so it is logical to use `String` type instead of `Vec<u8>`
- this type forcing you to allocate even when you don't need that
- at the end it is better to replace all `*_with_custom_entities` methods with the `*_with(expand_entity: impl Fn(&str)->&str)` method

Reply:
So basically something along these lines

```Rust
pub fn unescape(raw: &[u8]) -> Result<Cow<[u8]>, EscapeError> {
    do_unescape(raw, |_| None)
}

pub fn unescape_with<'a>(raw: &'a [u8], lookup_custom_entity: impl Fn(&str)->Option<&str>) {
     do_unescape(raw, lookup_custom_entity)
}

pub fn do_unescape<'a>(raw: &'a [u8], lookup_custom_entity: impl Fn(&str) -> Option<&str>) {
    // ....
    // ...
                let pat = &raw[start + 1..end];
                if pat.starts_with(b"#") {
                    let entity = &pat[1..]; // starts after the #
                    let codepoint = parse_number(entity, start..end)?;
                    push_utf8(unescaped, codepoint);
                } else if let Some(s) = named_entity(pat) {
                    unescaped.extend_from_slice(s.as_bytes());
                } else if let Some(value) = lookup_custom_entity(pat) {
                    unescaped.extend_from_slice(value.as_bytes());
                } else {
                    return Err(EscapeError::UnrecognizedSymbol(
                        start + 1..end,
                        String::from_utf8(pat.to_vec()),
                    ));
                }
     // ...
     // ...
}

fn main() {

   let custom_entities: HashMap<&'static str, &'static str> = HashMap::new();
   custom_entities.insert("anim", "lamb");

   let lookup_entity: |&str| -> Option<&str> = |ent| {
        custom_entities.get(ent)
   };

    println!("{}" unescape_with("mary had a little &anim;".as_bytes(), lookup_entity));
}

```

Reply:
I'm going to split this off into a separate PR